Introducing Python standard logging

Verequus opened the pull request:
Instead using stdout and stderr directly in the code (with or without Logger), all output is redirected over the standard Python logging API, which then allows to choose the output channels. Currently, only stdout and stderr are used. It is possible to add other channels if desired (like for allowing easier access to logs, if used as included library, once such functionality is provided).

It is also possible to add more logger instances using different names for every subfunctionality, which would allow to configure logging for each piece differently. That hasn't been done, but the minimum infrastructure to support this has been added.

Unfortunately, not all tests work. I couldn't figure out how to fix these, so this PR remains in draft mode.

Closes #312

latk commented:
Moving to standard logging sounds sensible. There are some caveats:

* Some output MUST go to stdout/stderr.
* Passing the logger object everywhere seems annoying.
* As currently written, this seems to complicate the tests.

## Some output must go to stdout/stderr

By default, all errors, warnings, and verbose-messages must go to stderr in order to conform to expectations about Unix-y tools. This can likely be achieved by configuring the logger appropriately.

There are a few outputs that also must go to stdout because they are essential output rather than messages:

* the summary-writer output
* the `--version` output
* the `--help` output (bypasses logging anyway)

It could make sense to always print these to stdout directly, without involving the logger.

## Passing the logger object everywhere seems annoying

The current `Logger` class exists so that it's not necessary to pass `options.verbose` around, and to provide basic utilities for formatting.

Real logging libraries (incl. Python's `logging`) typically use a singleton design where the logger can be retrieved using a global function, and the logger is configured once on program startup. So a typical Python source file would start with:

```python
import logging

LOG = logging.getLogger(__name__)
```

(in case the logger hierachy should follow the module hierarchy).

This PR adds `logger` arguments to a lot of functions. If we're already moving to standard logging, it would make sense to embrace this fully and to get rid of these explicit arguments. The formatting tasks of the existing Logger class can likely be moved to a logging-plugin. Unfortunately, the logging library design is so complicated that I wouldn't know how to do this.

Extra problem: some tests abuse logging for fault injection purposes. Untangling this might be tricky.

## Complications with testing

Right now, some tests seem to break. A lot of tests capture/redirect stderr+stdout in order to verify that a certain output was generated.

As we are using Pytest to run the tests, a lot of the `capsys` fixtures can probably be changed to the `caplog` fixture instead, see <https://docs.pytest.org/en/stable/logging.html>. But this might fail to detect actual regressions in the output.

It could also be that Pytest is currently conflicting with the logging. E.g. the tests like `test_empty_root` essentially runs `gcovr --root ''`. In reality, this works fine. In the tests, the capsys-captured stderr output seems to contain a `--- Logging error ---` which originates from within the logging module. Perhaps our `capture()` helper function that runs gcovr's main() needs to somehow isolate gcovr's logging from pytest's logging? I know far too little about the standard-library logging module to be of any help with debugging here.

Verequus commented:
### Some output MUST go to stdout/stderr.

Currently, the existing logger doesn't do that. Or rather, it entirely depends on the used function where the log ends up. It is only one endpoint. So anything which needs to go to both endpoints would need to call the logger twice already. Which isn't impacted in the first place due to switching to python logging as I implemented it. One can enhance the gcovr logger to provide a function which logs to both endpoints at once. Unless not using the gcovr logger class at all is desired, but that leaves the situation no worse than it is now.

For explicit stdout messages (the summary at least), I'd like to have them still at least in addition provide them in the logging system. In my project, we can't capture the output if it is only going to stdout/stderr. I think the easiest solution is to have to a secondary logger which is swallowing the output by default which stops having the output on stdout twice.

### Passing the logger object everywhere seems annoying.

I wasn't sure how to approach this change. I went with the current approach, because it was closer to what the code currently does. But I can try to get to the singleton approach instead. If you refer with formatting tasks to the pattern like "(ERROR) ", then it should be possible to automatically have them added in depending on the debug level. At least I've seen a similar example when looking for doing the proposed logging configuration.

In regards of BrokenLogger replacement: Patching the call to the log should work, but likely would require a way to trigger that only for DEBUG level. Maybe one could simply add an additional filter to a log handler instead. But that would require removing the additional filter afterwards. From a glance it seems the test can do the cleanup work as well.

### Complications with testing

Unnoticed regressions imply that certain output would never end up on the logging infrastructure. Is that actually possible? Maybe we need an exception handler to ensure capturing the exception messages, but otherwise where could this occur? Inside gcovr we have complete control over the output. And I guess that used libraries don't use stdout/stderr themselves.

In regards to pytest itself do logging, it should be possible to disable that completely. But before I investigate this, is the pytest log output merely an accidental catch?

Spacetown commented:
I'm with @latk to use the singleton object. We should use child loggers for each part of gcovr, removing the currently used arguments. The logging must be done to stderr. The format can be discussed or maybe configured by the user.
Writing output files to stdout should be done directly, bypassing the logger.

Why isn't it possible for you to capture stdout and stderr in your tests? The output shouldn't be added to the log. You can use a output file to get the data as file.

latk commented:
> Why isn't it possible for you to capture stdout and stderr in your tests? The output shouldn't be added to the log. You can use a output file to get the data as file.

In `test_args.py` we call the `main()` function directly (no separate process), and use Pytest's `capsys` fixtures to get stdout/stderr streams. This doesn't seem to properly affect logging though, and the output is garbled. Example output:

```none
_____________________________ test_invalid_objdir ______________________________

capsys = <_pytest.capture.CaptureFixture object at 0x7f8cb35dd080>

    def test_invalid_objdir(capsys):
        c = capture(capsys, ['--object-directory', 'not-existing-dir'])
        assert c.out == ''
>       assert c.err.startswith(
            '(ERROR) Bad --object-directory option.')
E       assert False
E        +  where False = <built-in method startswith of str object at 0x1b9f5a0>('(ERROR) Bad --object-directory option.')
E        +    where <built-in method startswith of str object at 0x1b9f5a0> = '--- Logging error ---\nTraceback (most recent call last):\n  File "/usr/lib/python3.6/logging/__init__.py", line 996,... does not exist.\'\nArguments: ()\n(ERROR) Bad --object-directory option.\n\tThe specified directory does not exist.\n'.startswith
E        +      where '--- Logging error ---\nTraceback (most recent call last):\n  File "/usr/lib/python3.6/logging/__init__.py", line 996,... does not exist.\'\nArguments: ()\n(ERROR) Bad --object-directory option.\n\tThe specified directory does not exist.\n' = <gcovr.tests.test_args.CaptureObject object at 0x7f8cb36b0320>.err

gcovr/tests/test_args.py:98: AssertionError
----------------------------- Captured stderr call -----------------------------
(ERROR) Bad --object-directory option.
        The specified directory does not exist.
(ERROR) Bad --object-directory option.
        The specified directory does not exist.
------------------------------ Captured log call -------------------------------
ERROR    gcovr.utils:utils.py:296 (ERROR) Bad --object-directory option.
        The specified directory does not exist.
```

Observations:

* Pytest captures log calls in addition to stdout/stderr.
* Pytest's captured stderr repeats the same message twice.
* The stderr captured via the `capsys` fixture in our `c` Capture object contains a logging error traceback message, suggesting an internal error in the logging library.

I think this points to a misconfiguration, but it could also be that Pytest is trying to be too clever and is getting in the way here. Using Pytest's `caplog` fixture may or may not help.

Spacetown commented:
> In my project, we can't capture the output if it is only going to stdout/stderr.

@latk I was wondering about this statement. I think our test can be adapted to work with the logger of python.

Verequus commented:
I tried to address to the raised objections. I've now placed logger creation in singleton objects where needed. I've gone with a fixed name to ensure that every file does access the same preconfigured logger. It is possible to change that name for different files, but I wasn't sure which strategy is intended to be used. But any new logger needs to be preconfigured in main() like the main one. It also means that as input you can't use additional arguments for formatting (not sure what those can be used for - Python documentation is completely lacking there). Instead, format strings take over the work. Which just moves the formatting before the log call and not during, so there is no practical issue with that change.

I remembered that you can patch objects in Python for UTs, so I replaced the BrokenLogger instance appropriately. Unfortunately, I'm not sure if that worked since the logging for tests still interferes. Replacing capsys with caplog everywhere merely increased failing tests, so that is not going to work.

I also changed the summary to have a parameter which if not given simply outputs to stdout as before. But if given, uses the logger. For whatever reason, a test breaks but I couldn't trace what the issue is here.

I hope that this overall change is closer to your view of how the code should look like. Still, the breaking tests are beyound my capability to fix.

Spacetown commented:
Oh, there is a problem with your updated branch. For me it seems that this is a merge mistake.
Also bak files from Notepad++ are added.

I've removed the last merge on your branch.

Spacetown commented on gcovr/__main__.py:
```diff
@@ -261,9 +264,9 @@ def main(args=None):
         ('--exclude-directories', options.exclude_dirs),
     ]:
-        logger.verbose_msg('Filters for {}: ({})', name, len(filters))
+        logger.debug(f'Filters for {name}: ({len(filters)})')
```
If strings are changed use double quote to make black happy.
```suggestion
        logger.debug(f"Filters for {name}: ({len(filters)})")
```

Spacetown commented on gcovr/__main__.py:
```diff
@@ -261,9 +264,9 @@ def main(args=None):
+        logger.debug(f'Filters for {name}: ({len(filters)})')
         for f in filters:
-            logger.verbose_msg('- {}', f)
+            logger.debug(f'- {f}')
```
```suggestion
            logger.debug(f" - {f}")
```

Spacetown commented on gcovr/configuration.py:
```diff
@@ -616,6 +616,12 @@ def merge_options_and_set_defaults(partial_namespaces, all_options=None):
         type=OutputOrDefault,
         default=None,
     ),
+    GcovrConfigOption(
+        "log_summary", ["--log_summary-uncovered"],
+        group="output_options",
+        help="Instead using stdout only, the summary uses logging as well.",
+        action="store_false",
+    ),
```
The summary is a output file and should not be logged.
```suggestion
```

Verequus replied:
I'm not able to access stdout in my project. How should I reroute this instead? Provide an output file instead?

Spacetown replied:
Why are you not able to access stdout? You can also redirect it to a file in the shell.

Verequus replied:
While in theory I can encapsulate the call of our python script with a redirection into a file, since we not just log lines to our log files, but to stdout in general, this redirection would capture the entire content of the dedicated log file in addition to the summary. Which means that piping the contents of the redirection file into the log file leads to needless duplication. Copying that redirection file as secondary output causes just confusion. Removing the duplication somehow makes more effort and is going against how we deal with logging in general. So in worst case we'd rather patch that part locally, if we upgrade later.

But I don't get your opposition to this approach. This is hidden behind a flag. And if it is active then you still get it in the console, since the logger is configured to output to stdout. In our case, we would be able to reconfigure the logger ourselves to use the same log file as the rest of the script. So what is a better solution that can satisfy us both?

Spacetown replied:
If the logger always goes to `stderr` a simply redirection should be enough.

latk replied:
Gcovr is a Unix-ish command line tool, which means that output will be generated on STDOUT and messages/warnings/errors are shown on STDERR. I'm fine with all the warnings being handled via standard logging, but I'm not OK with normal output of our reports taking this more fragile route. And the --print-summary output is effectively a kind of report.

I want to enable users to work effectively with gcovr. But I have to balance that against complexity and future maintenance effort. “This is hidden behind a flag” is not a strong argument for adding a feature. To me as a maintainer this sounds like “more complexity for marginal benefit”.

Eric Gunnerson has written a [great article about managing complexity on the C# compiler](https://docs.microsoft.com/en-us/archive/blogs/ericgu/minus-100-points). Core quotes:

> I've been spending a lot of time answering customer questions about why we do (or don't) have certain language features in C#. […]
>
> One of the big reasons we didn't do this is that it's really hard to remove complexity when you take a subtractive approach, as removing a feature in one area may not allow you to revisit low-level design decisions, nor will it allow you to remove complexity elsewhere, in places where it support the now-removed feature.
>
> So, we decided on the additive approach instead, and worked hard to keep the complexity down. One way to do that is through the concept of “minus 100 points”. Every feature starts out in the hole by 100 points, which means that it has to have a significant net positive effect on the overall package for it to make it into the language. Some features are okay features for a language to have, they just aren't quite good enough to make it into the language.

Here, we have the “--print-summary output should optionally be handled via logging” feature. It starts with -100 points. You're suggesting that it would make using gcovr in your specific environment easier where you cannot use tools that use the normal Unix conventions. That's maybe +1 point from my perspective, but still leaves -99 points.

This would also be a feature that boxes us into a certain design decision that we can't change later. Specifically, gcovr has not made and will never make any guarantees about how the output or messages arrive at STDOUT or STDERR. Whether we use `logging` or not is purely an implementation detail. Adding this option would change that from an implementation detail to something that's part of the public interface of this software. I'm very conscious about that in order to avoid [xkcd 1172](https://xkcd.com/1172/) situations.

I'm still very happy with the idea to use standard logging because the old `Logger` class makes the code unnecessarily complicated by having to pass around a logger object explicitly.

Verequus replied:
Ok, I'll remove that part then.

Spacetown commented on gcovr/decision_analysis.py:
```diff
@@ -14,6 +14,8 @@
     DecisionCoverageSwitch,
 )
 
+logger = logging.getLogger("gcovr")
```
```suggestion
logger = logging.getLogger(__name__)
```

Spacetown commented on gcovr/gcov.py:
```diff
@@ -15,46 +15,48 @@
 from .coverage import FileCoverage
 from .gcov_parser import parse_metadata, parse_coverage, ParserFlags
 
+logger = logging.getLogger("gcovr")
```
```suggestion
logger = logging.getLogger(__name__)
```

Spacetown commented on gcovr/gcov.py:
```diff
@@ -76,7 +78,7 @@ def find_datafiles(search_path, logger, exclude_dirs):
         if os.path.splitext(filename)[0] not in known_file_stems
     ]
-    logger.verbose_msg(
+    logger.debug(
```
Change also to f-string.

Verequus replied:
Done.

Spacetown commented on gcovr/gcov.py:
```diff
@@ -172,7 +169,7 @@ def guess_source_file_name(
         )
 
-    logger.verbose_msg(
+    logger.debug(
```
Change also to f-string.

Verequus replied:
Done.

Spacetown commented on gcovr/gcov.py:
```diff
@@ -341,7 +335,7 @@ def process_datafile(filename, covdata, options, toerase):
             return
 
-    logger.warn(
+    logger.warning(
```
Change also to f-string.

Verequus replied:
Done.

Spacetown commented on gcovr/gcov_parser.py:
```diff
@@ -50,7 +51,8 @@
 from .decision_analysis import DecisionParser
-from .utils import Logger
+
+logger = logging.getLogger("gcovr")
```
```suggestion
logger = logging.getLogger(__name__)
```

Spacetown commented on gcovr/gcov_parser.py:
```diff
@@ -480,25 +479,24 @@ def _report_lines_with_errors(
-        "\tat <https://github.com/gcovr/gcovr/issues>.",
-        source=context.filename,
-        lines="\n\t  ".join(lines),
+    lines_output = "\n\t  ".join(lines)
+    logger.warning(
+        f"Unrecognized GCOV output for {context.filename}\n"
+        f"\t  {lines_output}\n"
+        f"\tThis is indicative of a gcov output parse error.\n"
+        f"\tPlease report this to the gcovr developers\n"
+        f"\tat <https://github.com/gcovr/gcovr/issues>."
```
```suggestion
    logger.warning(
        f"Unrecognized GCOV output for {context.filename}\n"
        "\t  {'\n\t  '.join(lines)\n"
        "\tThis is indicative of a gcov output parse error.\n"
        "\tPlease report this to the gcovr developers\n"
        "\tat <https://github.com/gcovr/gcovr/issues>."
```

Verequus replied:
That change will not work. Using escape sequences inside a f-string at this place is not allowed according to Python spec. I tried that already.

Spacetown replied:
Ok. Please remove the f from the strings which don't contain a reference.

Verequus replied:
In my experience, I either assume that all string pieces are f-strings or none. In other words, this proofs against the bug, where someone adds some format string in the normal string pieces and forgets to change them into f-strings.

If you still insist, I will change it.

Spacetown replied:
There is a flake8 check which forbids f-strings where they aren't needed.

Verequus replied:
Done.

Spacetown commented on gcovr/utils.py:
```diff
@@ -17,11 +17,14 @@
 import sys
 from contextlib import contextmanager
 
+logger = logging.getLogger("gcovr")
```
```suggestion
logger = logging.getLogger(__name__)
```

Spacetown commented on gcovr/utils.py:
```diff
@@ -290,44 +293,44 @@ def match(self, path):
-        """Write a formatted error to STDERR.
+    # stderr warning configuration
+    stderr_warning_handler = logging.StreamHandler(sys.stderr)
+    warning_and_lower = MaximumApplicableLogLevelFilter(logging.WARNING)
+    stdout_handler.addFilter(warning_and_lower)
+    stderr_warning_handler.setFormatter(logging.Formatter("(WARNING) %(message)s"))
+    stderr_warning_handler.setLevel(logging.WARNING)
 
-        pattern: a str.format pattern
-        args, kwargs: str.format parameters
-        """
-        pattern = "(ERROR) " + pattern + "\n"
-        sys.stderr.write(pattern.format(*args, **kwargs))
+    # stderr warning configuration
+    stderr_error_handler = logging.StreamHandler(sys.stderr)
+    stderr_error_handler.setFormatter(logging.Formatter("(ERROR) %(message)s"))
+    stderr_error_handler.setLevel(logging.ERROR)
 
-    def msg(self, pattern, *args, **kwargs):
-        """Write a formatted message to STDOUT.
+    logger.addHandler(stdout_handler)
+    logger.addHandler(stderr_warning_handler)
+    logger.addHandler(stderr_error_handler)
 
-        pattern: a str.format pattern
-        args, kwargs: str.format arguments
-        """
-        pattern = pattern + "\n"
-        sys.stdout.write(pattern.format(*args, **kwargs))
+    sys.excepthook = exception_hook
 
-    def verbose_msg(self, pattern, *args, **kwargs):
-        """Write a formatted message to STDOUT if in verbose mode.
 
-        see: self.msg()
-        """
-        if self.verbose:
-            self.msg(pattern, *args, **kwargs)
+def exception_hook(exc_type, exc_value, exc_traceback) -> None:
+    logging.exception("Uncaught EXCEPTION", exc_info=(exc_type, exc_value, exc_traceback))
```
Is this complicated configuration really needed? Logging is always done to `stderr` and the message can be configured with the format:
```python
"(%(levelname)s) %(message)s"
```

Verequus replied:
This logging configuration replicates the old Logger class functionality completely. Look at the actual implementation. So splitting between `stderr` and `stdout` is required, as well that different log levels use a different message format. Unless you want to change the tests as well, this has to be done like this. At least as first step, to avoid too many changes at ones.

Otherwise, `stderr` and  `"(%(levelname)s) %(message)s"` would simplify things a bit.

Spacetown replied:
> It could make sense to always print these to stdout directly, without involving the logger.

Configur logger in main function to always print to `stderr`. The only change here is that infos get a `INFO` prepended.

Verequus replied:
Changed this to always log to `stderr` and using `"(%(levelname)s) %(message)s"` to always add the level in the output.

Spacetown commented on gcovr/writer/html.py:
```diff
@@ -15,14 +15,16 @@
-from ..utils import realpath, commonpath, sort_coverage, calculate_coverage, Logger, open_text_for_writing
+from ..utils import realpath, commonpath, sort_coverage, calculate_coverage, open_text_for_writing
+
+logger = logging.getLogger("gcovr")
```
```suggestion
logger = logging.getLogger(__name__)
```

Spacetown commented on gcovr/writer/summary.py:
```diff
@@ -15,31 +15,38 @@
-    sys.stdout.write(lines_out)
-    sys.stdout.write(functions_out)
-    sys.stdout.write(branches_out)
+    if log_summary:
+        logger.info(lines_out)
+        logger.info(functions_out)
+        logger.info(branches_out)
+    else:
+        sys.stdout.write(lines_out + '\n')
+        sys.stdout.write(functions_out + '\n')
+        sys.stdout.write(branches_out + '\n')
```
As written above, this is a output file.
```suggestion
    sys.stdout.write(lines_out + '\n')
    sys.stdout.write(functions_out + '\n')
    sys.stdout.write(branches_out + '\n')
```

Spacetown commented on gcovr/writer/summary.py:
```diff
@@ -15,31 +15,38 @@
 
-def print_summary(covdata):
-    '''Print a small report to the standard output.
+def print_summary(covdata, log_summary):
+    '''Print a small report to the standard output or to the logger in addition.
```
As written above, this is a output file.
```suggestion
def print_summary(covdata):
    '''Print a small report to the standard output.
```

Spacetown reviewed (changes requested):
As already mentioned by @latk must the summary output always go to STDOUT.
It's better to use the hierarchical names via `__name__` than hardcode the string in each module.

Verequus commented:
> Oh, there is a problem with your updated branch. For me it seems that this is a merge mistake.
> Also bak files from Notepad++ are added.

> I've removed the last merge on your branch.

Now I'm not sure exactly I need to aside of re-removing the bak files (pretty sure I had that in the merge). How can I see what changes you actually removed?

Verequus commented on gcovr/__main__.py:
```diff
@@ -47,10 +47,13 @@
 from .writer.coveralls import print_coveralls_report
 
 
+logger = logging.getLogger('gcovr')
```
The issue with that is the configuration of that logger. While I can call the configuration function in every file separately, I don't know how to access the verbose option from the main function. It would either lead to have DEBUG messages always displayed or never. If I do not fix the name, I can't access the logger reliably from the outside.

How to solve the problem?

Spacetown replied:
This should be the name of the module and this is fix.

Verequus replied:
I'm sorry, but I'm not sure how to read your reply. What do you suggest exactly?

Spacetown replied:
Ok, I got it. In the main script `__name__` isn't `gcovr`. So here it can stay hardcoded but with double quotes.

Verequus replied:
Quotes changed.

Verequus commented:
I've updated the code to remove erroneous usage of logger and a number of other mistakes buried in the test output. I also managed to fix most tests which did break. Or at least, those which were definitely broken. I'm not sure if the other tests do work or appear to work only because the output by chance matches the checks. I didn't have time to figure that part out.

In any case, only 3 tests still complain and I have no clue how to fix them.

Verequus commented:
No chance to look at this yet? Just wondering about the status.

Spacetown commented:
I fixed the test:
- `logger.info` instead of `print`was used for the output of `--version`.
- The doctest for `_ExclusionRangeWarnings` fails because the add of the handler didn't work as expected. Now the `caplog` fixture is used to check the messages.
- In `test_filter_backslashes_are_detected` the messages where compared with a trailing linebreak.

I try to check the code tomorrow.

Spacetown commented:
@Verequus The tests are fixed now. Please can you double check the changes?

Verequus commented:
@Spacetown I noticed an inconsistency with my code changes. You used:

>     message = c.record_tuples[0]
>     assert message[1] == logging.WARNING
>     assert message[2] == "HTML output skipped - consider providing an output file with `--html=OUTPUT`."
>     message = c.record_tuples[1]
>     assert message[1] == logging.WARNING
>     assert message[2] == "Sonarqube output skipped - consider providing an output file with `--sonarqube=OUTPUT`."
>     message = c.record_tuples[2]
>     assert message[1] == logging.WARNING
>     assert message[2] == "Coveralls output skipped - consider providing an output file with `--coveralls=OUTPUT`."

So you reused message. I would have written this instead:

>     message0 = c.record_tuples[0]
>     assert message0[1] == logging.WARNING
>     assert message0[2] == "HTML output skipped - consider providing an output file with `--html=OUTPUT`."
>     message1 = c.record_tuples[1]
>     assert message1[1] == logging.WARNING
>     assert message1[2] == "Sonarqube output skipped - consider providing an output file with `--sonarqube=OUTPUT`."
>     message2 = c.record_tuples[2]
>     assert message2[1] == logging.WARNING
>     assert message2[2] == "Coveralls output skipped - consider providing an output file with `--coveralls=OUTPUT`."

So numbering the variables. I don't have any particular preferences (originally expected to have to reuse the variables more), but it should only done one way. What do you want?

Otherwise I'm ok. Just wondering, why are double quotes preferred in the first place?

Spacetown commented:
I also have no preference. The double quotes are required by black. Single quotes shall only be used if you use double quotes inside the string.

Verequus commented:
I've chosen your way since it makes it for you easier to maintain the code. So I'd say you can merge this, if you accept the codecov failures.

Spacetown reviewed (approved):
LGFM

latk commented on gcovr/gcov_parser.py:
```diff
@@ -867,71 +857,56 @@ class _ExclusionRangeWarnings:
-    (WARNING) The coverage exclusion region start flag GCOVR_EXCL_START
+    30: GCOVR_EXCL_LINE found on line 4 in excluded region started on line 3, when processing example.cpp.
+    30: GCOVR_EXCL_START found on line 3 was terminated by GCOV_EXCL_STOP on line 5, when processing example.cpp.
+    30: The coverage exclusion region start flag {start}
```
The change on this line looks like a regression – the placeholder `{start}` should have been replaced by the variable.

latk commented on gcovr/gcov_parser.py:
```diff
@@ -867,71 +857,56 @@ class _ExclusionRangeWarnings:
         """warn that a region was started but not closed"""
-        self.logger.warn(
+        logger.warning(
             "The coverage exclusion region start flag {start}\n"
```
Ah, this is the line with the placeholder – it should now be an `f""`-string.

latk commented on gcovr/tests/test_args.py:
```diff
@@ -49,10 +50,32 @@ def capture(capsys, args, other_ex=()):
+        self.exception = exception
+
+
+# The capture method calls the main method and captures its output/error
+# streams and exit code
```
This comment wants to be a docstring :)

Spacetown replied:
Also for the function `coverage`.

latk commented on gcovr/writer/summary.py:
```diff
@@ -15,11 +15,13 @@
 
 from ..utils import get_global_stats
 
+logger = logging.getLogger("gcovr")
```
unused

latk reviewed (commented):
Thank you @Verequus and @Spacetown for working on this! Glad to see the old Logger go.

I think I found a tiny regression in an exclusion marker warning, which we should fix (in a separate PR) before shipping a new version.

I'm glad that a way was found to make the tests run, even though the changes shifted the meanings of the tests. Previously, they also asserted that STDOUT was empty, ensuring that there were no stray print() calls for logging. This is not something that has to be fixed, it's just something I'm thinking loudly about in case a better approach becomes apparent.

Spacetown commented:
I'll create a PR for this.

Verequus commented:
@latk You're welcome!

In regards of stray prints: Not sure if you can have both caplog and capsys. If you do then it would be easy to look for the text string being empty in every test. Otherwise, one would need to have some code check for the number of prints in files. Any additions would be discovered while scanning over it.

Spacetown commented:
I also checked this but found no solution. You've to decide if you want to use `capsys` OR `caplog`.
One solution can be to use a flake8 plugin to forbid several functions.

Verequus commented:
@Spacetown I occurred to me, if you duplicate the caplog testcases as capsys testcases and temporarily change the logging to not output anything to stderr, you can test if there is any other output.

Spacetown commented:
@latk What do you think about this? The problem would be that this is only for a specific test where a flake8 test scans all branches in the code.

latk commented:
I think I would be clever enough to write a flake8 plugin that traverses the AST of a file to find uses of illegal functions. But I know I won't find the time for it :)

However, there has to already be some feature of flake8 or pylint that lets us blocklist certain functions. Maybe adding a Pylint configuration could make sense. We'd have to disable some of the annoying lints (especially those relating to missing docstrings) but its `bad-functions` config option could work. The Pylint docs are super bad but the [source code](https://github.com/PyCQA/pylint/blob/v2.12.2/pylint/extensions/bad_builtin.py) for this checker explains what's going on.

Lints and tests have different tradeoffs. Testing shows the absence of certain behaviour, but only in those examples. Linting shows the absence of certain coding patterns, but doesn't guarantee any particular behaviour. But since linting is easier for us to use in this case, it's entirely sufficient.

Spacetown commented:
I'll check this. The forbidden function at the moment would be `print` which will be turned of for the version output, the summary and the tests.

Spacetown commented:
I found https://github.com/JBKahn/flake8-print which does exactly what we want.
